src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.ValidateSignature.cs

```diff
@@ -66,57 +65,40 @@ internal static ValidationResult<SecurityKey> ValidateSignature(
                 resolvedKey = SamlTokenUtilities.ResolveTokenSigningKey(samlToken.Assertion.Signature.KeyInfo, validationParameters);
             }
 
-            if (resolvedKey is null)
+            bool canMatchKey = samlToken.Assertion.Signature.KeyInfo != null;
+            List<ValidationError>? errors = null;
+            ValidationError? error = null;
+            StringBuilder? keysAttempted = null;
+
+            if (resolvedKey is not null)
             {
-                if (validationParameters.TryAllIssuerSigningKeys)
-                    keys = validationParameters.IssuerSigningKeys;
+                keyMatched = true;
+                var result = ValidateSignatureUsingKey(resolvedKey, samlToken, validationParameters, callContext);
+                if (result.IsValid)
+                    return result;
+
+                error = result.UnwrapError();
             }
             else
             {
-                keys = [resolvedKey];
-                keyMatched = true;
+                if (validationParameters.TryAllIssuerSigningKeys)
+                    keys = validationParameters.IssuerSigningKeys;
             }
 
-            bool canMatchKey = samlToken.Assertion.Signature.KeyInfo != null;
-            List<ValidationError> errors = new();
-            StringBuilder keysAttempted = new();
-
             if (keys is not null)
             {
+                // Control reaches here only if the key could not be resolved and TryAllIssuerSigningKeys is set to true.
+                // We try all the keys in the list and return the first valid key. This is the degenerate case.
```

Comment:
> This is the degenerate case.

**Question (non-blocking)** What does "degenerate case" mean in this context? Are there other cases that can be called out in comments to clarify this?

Reply:
I originally borrowed the term from our JWT handler. I read it as a bit of an edge case where we are not focused on obtaining the best performance, as if that were the case, we would have gone with the alternative/trivial case.

In this context, it means that the application was configured to say "if you can't resolve the key, here's potentially a bunch of others you can try, even though that may take time", by enabling `TryAllIssuerKeys` and receiving multiple keys to try.

The expected most common scenario is that the token has a key ID identifying the key it was signed with, we either received from the user or from the configuration a key with that same ID, we match them, verify the signature and either succeed or fail in a single attempt.